kive/container/tests.py

```diff
@@ -3077,6 +3080,104 @@ def test_rerun_input_missing(self):
 
         self.assert_run_fails(run4, 'RuntimeError: Inputs missing from reruns')
 
+    def test_fixed_argument_formatting(self):
+        app = ContainerApp.objects.first()
+        run = app.runs.first()
+        args = list(app.arguments.all())
+        command = runcontainer.Command.build_command(run)
+
+        expected = []
+        for arg in args:
+            if arg.type == ContainerArgument.INPUT:
+                expected.append("/mnt/input/{}".format(arg.name))
+            else:
+                expected.append("/mnt/output/{}".format(arg.name))
+        self.assertEqual(len(args), 2, "Expected run to have two arguments")
+        self.assertEqual(expected, command[-len(expected):])
+
+    def test_full_argument_formatting(self):
+        # Set up test data
+        # NOTE(nknight): The order of these specs matters; it matches the order that the
+        # associated ContainerDataset objects would be returned from the database.
+        argspecs = [
+            {
+                "name": "positional_input",
+                "position": 0,
+                "type": ContainerArgument.INPUT,
+            },
+            {
+                "name": "positional_output",
+                "position": 2,
+                "type": ContainerArgument.OUTPUT
+            },
+            {
+                "name": "optional_input",
+                "position": None,
+                "type": ContainerArgument.INPUT,
+            },
+            # TODO(nknight): Add spec for an optional multiple input
+            # TODO(nknight): Add spec for a positional output directory
+        ]
+
+        app = ContainerApp(
+            container=Container.objects.get(pk=1),
+        )
+        app.save()
+
+        args = [ContainerArgument(app=app, **spec) for spec in argspecs]
+        for arg in args:
+            arg.save()
+
+        datasets = []
+        for arg in (a for a in args if a.type == ContainerArgument.INPUT):
+            dataset = unittest.mock.Mock()
+            dataset.argument = arg
+            dataset.name = arg.name
+            datasets.append(dataset)
+
+        mock_run = unittest.mock.Mock()
+        mock_run.app = app
+        mock_run.container.file.path = "mock_container_path"
+        mock_run.full_sandbox_path = "mock_sandbox_path"
+        mock_run.datasets.all = unittest.mock.Mock(return_value=iter(datasets))
+
+        # Exercise method under test
+        full_command = runcontainer.Command.build_command(mock_run)
+
+        # Catch changes to the command list's length
+        self.assertNotIn("--app", full_command, "Unexpected '--app' name in command")
+        command_args = full_command[7:]
+
+        # Parse optional and positional args
```

Comment:
Is it useful to pull the arg list apart? I think the test would be easier to understand if you just asserted equality on the whole list.  You could add comments to the expected list if you want to explain what each part is.

Reply:
This test originally (and I hope in the future) had several keyword-style arguments whose order doesn't matter. Parsed them out so that they could be compared as dictionaries in case their order changed (for whatever reason).

If you think it would be clearer, I could factor the parsing code into its own method so that it doesn't break up the test code quite as much.

Reply:
The nice thing about dictionaries with stable ordering is that the order of arguments shouldn't change. I suggest you simplify the test until we find a reason that we need it more flexible.

Reply:
Simplified code is pushed.